File: observer-ts/src/otlp/convert.ts
// Convert OTLP JSON structures to store types.
// Handles both JSON-encoded OTLP and decoded protobuf (same shape after decode).

import type { Span, MetricDataPoint, LogRecord, Resource, Scope, SpanEvent, SpanLink } from "../store/types.ts";

// --- OTLP JSON shapes (subset needed for conversion) ---

type OtlpAnyValue =
  | { stringValue: string }
  | { intValue: string | number }
  | { doubleValue: number }
  | { boolValue: boolean }
  | { arrayValue: { values: OtlpAnyValue[] } }
  | { kvlistValue: { values: OtlpKeyValue[] } }
  | { bytesValue: string };

type OtlpKeyValue = {
  key: string;
  value: OtlpAnyValue;
};

type OtlpResource = {
  attributes?: OtlpKeyValue[];
  droppedAttributesCount?: number;
};

type OtlpScope = {
  name?: string;
  version?: string;
  attributes?: OtlpKeyValue[];
};

// --- Traces ---

type OtlpSpan = {
  traceId: string;
  spanId: string;
  parentSpanId?: string;
  name: string;
  kind?: number | string;
  startTimeUnixNano: string;
  endTimeUnixNano: string;
  attributes?: OtlpKeyValue[];
  events?: OtlpSpanEvent[];
  links?: OtlpSpanLink[];
  status?: { code?: number | string; message?: string };
};

type OtlpSpanEvent = {
  name: string;
  timeUnixNano: string;
  attributes?: OtlpKeyValue[];
};

type OtlpSpanLink = {
  traceId: string;
  spanId: string;
  attributes?: OtlpKeyValue[];
};

type OtlpScopeSpans = {
  scope?: OtlpScope;
  spans?: OtlpSpan[];
  schemaUrl?: string;
};

type OtlpResourceSpans = {
  resource?: OtlpResource;
  scopeSpans?: OtlpScopeSpans[];
  schemaUrl?: string;
};

export type OtlpTracesPayload = {
  resourceSpans?: OtlpResourceSpans[];
};

// --- Metrics ---

type OtlpNumberDataPoint = {
  attributes?: OtlpKeyValue[];
  startTimeUnixNano?: string;
  timeUnixNano: string;
  asInt?: string | number;
  asDouble?: number;
  flags?: number;
};

type OtlpHistogramDataPoint = {
  attributes?: OtlpKeyValue[];
  startTimeUnixNano?: string;
  timeUnixNano: string;
  count?: string | number;
  sum?: number;
  min?: number;
  max?: number;
  bucketCounts?: (string | number)[];
  explicitBounds?: number[];
  flags?: number;
};

type OtlpSummaryDataPoint = {
  attributes?: OtlpKeyValue[];
  startTimeUnixNano?: string;
  timeUnixNano: string;
  count?: string | number;
  sum?: number;
  quantileValues?: { quantile: number; value: number }[];
  flags?: number;
};

type OtlpMetric = {
  name: string;
  description?: string;
  unit?: string;
  gauge?: { dataPoints?: OtlpNumberDataPoint[] };
  sum?: { dataPoints?: OtlpNumberDataPoint[]; isMonotonic?: boolean; aggregationTemporality?: number | string };
  histogram?: { dataPoints?: OtlpHistogramDataPoint[]; aggregationTemporality?: number | string };
  summary?: { dataPoints?: OtlpSummaryDataPoint[] };
  exponentialHistogram?: { dataPoints?: OtlpHistogramDataPoint[]; aggregationTemporality?: number | string };
};

type OtlpScopeMetrics = {
  scope?: OtlpScope;
  metrics?: OtlpMetric[];
  schemaUrl?: string;
};

type OtlpResourceMetrics = {
  resource?: OtlpResource;
  scopeMetrics?: OtlpScopeMetrics[];
  schemaUrl?: string;
};

export type OtlpMetricsPayload = {
  resourceMetrics?: OtlpResourceMetrics[];
};

// --- Logs ---

type OtlpLogRecord = {
  timeUnixNano?: string;
  observedTimeUnixNano?: string;
  severityNumber?: number;
  severityText?: string;
  body?: OtlpAnyValue;
  attributes?: OtlpKeyValue[];
  traceId?: string;
  spanId?: string;
  flags?: number;
};

type OtlpScopeLogs = {
  scope?: OtlpScope;
  logRecords?: OtlpLogRecord[];
  schemaUrl?: string;
};

type OtlpResourceLogs = {
  resource?: OtlpResource;
  scopeLogs?: OtlpScopeLogs[];
  schemaUrl?: string;
};

export type OtlpLogsPayload = {
  resourceLogs?: OtlpResourceLogs[];
};

// ─── Converters ────────────────────────────────────────

export function convertTraces(payload: OtlpTracesPayload): Span[] {
  const spans: Span[] = [];
  for (const rs of payload.resourceSpans ?? []) {
    const resource = convertResource(rs.resource);
    for (const ss of rs.scopeSpans ?? []) {
      const scope = convertScope(ss.scope);
      for (const sp of ss.spans ?? []) {
        const startNano = sp.startTimeUnixNano || "0";
        const endNano = sp.endTimeUnixNano || "0";
        const durationMs = (Number(endNano) - Number(startNano)) / 1_000_000;

        spans.push({
          traceId: decodeHexId(sp.traceId),
          spanId: decodeHexId(sp.spanId),
          parentSpanId: sp.parentSpanId ? decodeHexId(sp.parentSpanId) : undefined,
          name: sp.name,
          kind: spanKindToString(sp.kind),
          startTimeUnixNano: startNano,
          endTimeUnixNano: endNano,
          durationMs,
          status: {
            code: statusCodeToString(sp.status?.code),
            message: sp.status?.message || undefined,
          },
          attributes: convertAttributes(sp.attributes),
          events: (sp.events ?? []).map(convertEvent),
          links: (sp.links ?? []).map(convertLink),
          resource,
          scope,
        });
      }
    }
  }
  return spans;
}

export function convertMetrics(payload: OtlpMetricsPayload): MetricDataPoint[] {
  const points: MetricDataPoint[] = [];
  for (const rm of payload.resourceMetrics ?? []) {
    const resource = convertResource(rm.resource);
    for (const sm of rm.scopeMetrics ?? []) {
      const scope = convertScope(sm.scope);
      for (const m of sm.metrics ?? []) {
        if (m.gauge?.dataPoints) {
          for (const dp of m.gauge.dataPoints) {
            points.push(convertNumberPoint(m, dp, "gauge", resource, scope));
          }
        }
        if (m.sum?.dataPoints) {
          for (const dp of m.sum.dataPoints) {
            const p = convertNumberPoint(m, dp, "sum", resource, scope);
            p.isMonotonic = m.sum!.isMonotonic;
            p.temporality = temporalityToString(m.sum!.aggregationTemporality);
            points.push(p);
          }
        }
        if (m.histogram?.dataPoints) {
          for (const dp of m.histogram.dataPoints) {
            points.push(convertHistogramPoint(m, dp, "histogram", resource, scope, m.histogram!.aggregationTemporality));
          }
        }
        if (m.summary?.dataPoints) {
          for (const dp of m.summary.dataPoints) {
            points.push(convertSummaryPoint(m, dp, resource, scope));
          }
        }
        if (m.exponentialHistogram?.dataPoints) {
          for (const dp of m.exponentialHistogram.dataPoints) {
            points.push(convertHistogramPoint(m, dp, "exponential_histogram", resource, scope, m.exponentialHistogram!.aggregationTemporality));
          }
        }
      }
    }
  }
  return points;
}

export function convertLogs(payload: OtlpLogsPayload): LogRecord[] {
  const logs: LogRecord[] = [];
  for (const rl of payload.resourceLogs ?? []) {
    const resource = convertResource(rl.resource);
    for (const sl of rl.scopeLogs ?? []) {
      const scope = convertScope(sl.scope);
      for (const lr of sl.logRecords ?? []) {
        logs.push({
          timeUnixNano: lr.timeUnixNano || lr.observedTimeUnixNano || "0",
          severityNumber: lr.severityNumber,
          severityText: lr.severityText || undefined,
          body: anyValueToString(lr.body),
          attributes: convertAttributes(lr.attributes),
          traceId: lr.traceId ? decodeHexId(lr.traceId) : undefined,
          spanId: lr.spanId ? decodeHexId(lr.spanId) : undefined,
          resource,
          scope,
        });
      }
    }
  }
  return logs;
}

// ─── Helpers ───────────────────────────────────────────

function convertResource(r?: OtlpResource): Resource {
  const attrs = convertAttributes(r?.attributes);
  const serviceName = attrs["service.name"] as string | undefined;
  return {
    serviceName: serviceName || undefined,
    attributes: attrs,
  };
}

function convertScope(s?: OtlpScope): Scope {
  return {
    name: s?.name ?? "",
    version: s?.version || undefined,
  };
}

function convertAttributes(kvs?: OtlpKeyValue[]): Record<string, unknown> {
  if (!kvs) return {};
  const attrs: Record<string, unknown> = {};
  for (const kv of kvs) {
    attrs[kv.key] = anyValueToJs(kv.value);
  }
  return attrs;
}

function anyValueToJs(v: OtlpAnyValue | undefined): unknown {
  if (!v) return undefined;
  if ("stringValue" in v) return v.stringValue;
  if ("intValue" in v) return typeof v.intValue === "string" ? parseInt(v.intValue, 10) : v.intValue;
  if ("doubleValue" in v) return v.doubleValue;
  if ("boolValue" in v) return v.boolValue;
  if ("arrayValue" in v) return (v.arrayValue.values ?? []).map(anyValueToJs);
  if ("kvlistValue" in v) {
    const obj: Record<string, unknown> = {};
    for (const kv of v.kvlistValue.values ?? []) {
      obj[kv.key] = anyValueToJs(kv.value);
    }
    return obj;
  }
  if ("bytesValue" in v) return v.bytesValue;
  return undefined;
}

function anyValueToString(v?: OtlpAnyValue): string {
  if (!v) return "";
  const js = anyValueToJs(v);
  return typeof js === "string" ? js : JSON.stringify(js);
}

function convertEvent(e: OtlpSpanEvent): SpanEvent {
  return {
    name: e.name,
    timeUnixNano: e.timeUnixNano || "0",
    attributes: convertAttributes(e.attributes),
  };
}

function convertLink(l: OtlpSpanLink): SpanLink {
  return {
    traceId: decodeHexId(l.traceId),
    spanId: decodeHexId(l.spanId),
    attributes: convertAttributes(l.attributes),
  };
}

function convertNumberPoint(
  m: OtlpMetric,
  dp: OtlpNumberDataPoint,
  type: string,
  resource: Resource,
  scope: Scope,
): MetricDataPoint {
  return {
    name: m.name,
    description: m.description || undefined,
    unit: m.unit || undefined,
    type,
    timeUnixNano: dp.timeUnixNano,
    startTimeUnixNano: dp.startTimeUnixNano || undefined,
    attributes: convertAttributes(dp.attributes),
    resource,
    scope,
    flags: dp.flags,
    value: dp.asDouble ?? (dp.asInt != null ? Number(dp.asInt) : undefined),
  };
}

function convertHistogramPoint(
  m: OtlpMetric,
  dp: OtlpHistogramDataPoint,
  type: string,
  resource: Resource,
  scope: Scope,
  temporality?: number | string,
): MetricDataPoint {
  return {
    name: m.name,
    description: m.description || undefined,
    unit: m.unit || undefined,
    type,
    timeUnixNano: dp.timeUnixNano,
    startTimeUnixNano: dp.startTimeUnixNano || undefined,
    attributes: convertAttributes(dp.attributes),
    resource,
    scope,
    flags: dp.flags,
    count: dp.count != null ? Number(dp.count) : undefined,
    sum: dp.sum,
    min: dp.min,
    max: dp.max,
    bucketCounts: dp.bucketCounts?.map(Number),
    explicitBounds: dp.explicitBounds,
    temporality: temporalityToString(temporality),
  };
}

function convertSummaryPoint(
  m: OtlpMetric,
  dp: OtlpSummaryDataPoint,
  resource: Resource,
  scope: Scope,
): MetricDataPoint {
  return {
    name: m.name,
    description: m.description || undefined,
    unit: m.unit || undefined,
    type: "summary",
    timeUnixNano: dp.timeUnixNano,
    startTimeUnixNano: dp.startTimeUnixNano || undefined,
    attributes: convertAttributes(dp.attributes),
    resource,
    scope,
    flags: dp.flags,
    count: dp.count != null ? Number(dp.count) : undefined,
    sum: dp.sum,
    quantiles: dp.quantileValues,
  };
}

// ─── Enum conversions ──────────────────────────────────

const SPAN_KIND_MAP: Record<number, string> = {
  0: "UNSPECIFIED",
  1: "INTERNAL",
  2: "SERVER",
  3: "CLIENT",
  4: "PRODUCER",
  5: "CONSUMER",
};

function spanKindToString(kind?: number | string): string {
  if (typeof kind === "string") {
    // Already a string (e.g. "SPAN_KIND_SERVER" or "SERVER").
    return kind.replace("SPAN_KIND_", "");
  }
  return SPAN_KIND_MAP[kind ?? 0] ?? "UNSPECIFIED";
}

function statusCodeToString(code?: number | string): string {
  if (typeof code === "string") {
    return code.replace("STATUS_CODE_", "");
  }
  switch (code) {
    case 0: return "UNSET";
    case 1: return "OK";
    case 2: return "ERROR";
    default: return "UNSET";
  }
}

function temporalityToString(t?: number | string): string | undefined {
  if (typeof t === "string") return t;
  switch (t) {
    case 1: return "delta";
    case 2: return "cumulative";
    default: return undefined;
  }
}

/** Decode hex or base64 IDs to lowercase hex. */
function decodeHexId(id: string): string {
  if (!id) return "";
  // OTLP JSON uses base64 for IDs; OTLP HTTP JSON uses hex.
  // If it looks like hex already (16 or 32 hex chars), return lowercase.
  if (/^[0-9a-fA-F]+$/.test(id)) return id.toLowerCase();
  // Try base64 decode only if it looks like valid base64 (correct length, charset).
  if (/^[A-Za-z0-9+/]+=*$/.test(id) && id.length >= 12) {
    try {
      const bytes = Buffer.from(id, "base64");
      if (bytes.length === 8 || bytes.length === 16) {
        return bytes.toString("hex");
      }
    } catch {
      // Fall through.
    }
  }
  // Return as-is for non-standard IDs.
  return id;
}

File: observer/internal/otlp/convert.go
package otlp

import (
	"encoding/hex"

	"github.com/signalfx/obstudio/observer/internal/store"
	"go.opentelemetry.io/collector/pdata/pcommon"
	"go.opentelemetry.io/collector/pdata/plog"
	"go.opentelemetry.io/collector/pdata/pmetric"
	"go.opentelemetry.io/collector/pdata/ptrace"
)

func ConvertTraces(td ptrace.Traces) []store.Span {
	var out []store.Span
	rss := td.ResourceSpans()
	for i := 0; i < rss.Len(); i++ {
		rs := rss.At(i)
		res := convertResource(rs.Resource(), rs.SchemaUrl())
		sss := rs.ScopeSpans()
		for j := 0; j < sss.Len(); j++ {
			ss := sss.At(j)
			scope := convertScope(ss.Scope(), ss.SchemaUrl())
			spans := ss.Spans()
			for k := 0; k < spans.Len(); k++ {
				sp := spans.At(k)
				start := sp.StartTimestamp().AsTime()
				end := sp.EndTimestamp().AsTime()
				out = append(out, store.Span{
					TraceID:      traceIDStr(sp.TraceID()),
					SpanID:       spanIDStr(sp.SpanID()),
					ParentSpanID: spanIDStr(sp.ParentSpanID()),
					Name:         sp.Name(),
					Kind:         spanKindString(sp.Kind()),
					StartTime:    start,
					EndTime:      end,
					DurationMs:   float64(end.Sub(start).Microseconds()) / 1000.0,
					Status:       convertSpanStatus(sp.Status()),
					Attributes:   mapToGoMap(sp.Attributes()),
					Events:       convertEvents(sp.Events()),
					Links:        convertLinks(sp.Links()),
					Resource:     res,
					Scope:        scope,
				})
			}
		}
	}
	return out
}

func ConvertMetrics(md pmetric.Metrics) []store.MetricDataPoint {
	var out []store.MetricDataPoint
	rms := md.ResourceMetrics()
	for i := 0; i < rms.Len(); i++ {
		rm := rms.At(i)
		res := convertResource(rm.Resource(), rm.SchemaUrl())
		sms := rm.ScopeMetrics()
		for j := 0; j < sms.Len(); j++ {
			sm := sms.At(j)
			scope := convertScope(sm.Scope(), sm.SchemaUrl())
			ms := sm.Metrics()
			for k := 0; k < ms.Len(); k++ {
				out = append(out, convertMetric(ms.At(k), res, scope)...)
			}
		}
	}
	return out
}

func ConvertLogs(ld plog.Logs) []store.LogRecord {
	var out []store.LogRecord
	rls := ld.ResourceLogs()
	for i := 0; i < rls.Len(); i++ {
		rl := rls.At(i)
		res := convertResource(rl.Resource(), rl.SchemaUrl())
		sls := rl.ScopeLogs()
		for j := 0; j < sls.Len(); j++ {
			sl := sls.At(j)
			scope := convertScope(sl.Scope(), sl.SchemaUrl())
			lrs := sl.LogRecords()
			for k := 0; k < lrs.Len(); k++ {
				lr := lrs.At(k)
				out = append(out, store.LogRecord{
					Timestamp:      lr.Timestamp().AsTime(),
					SeverityNumber: int32(lr.SeverityNumber()),
					SeverityText:   lr.SeverityText(),
					Body:           valueString(lr.Body()),
					Attributes:     mapToGoMap(lr.Attributes()),
					TraceID:        traceIDStr(lr.TraceID()),
					SpanID:         spanIDStr(lr.SpanID()),
					Resource:       res,
					Scope:          scope,
				})
			}
		}
	}
	return out
}

func convertMetric(m pmetric.Metric, res store.Resource, scope store.Scope) []store.MetricDataPoint {
	name := m.Name()
	desc := m.Description()
	unit := m.Unit()

	switch m.Type() {
	case pmetric.MetricTypeGauge:
		return convertNumberDataPoints(m.Gauge().DataPoints(), name, desc, unit, "gauge", "", false, res, scope)
	case pmetric.MetricTypeSum:
		s := m.Sum()
		return convertNumberDataPoints(s.DataPoints(), name, desc, unit, "sum", temporalityString(s.AggregationTemporality()), s.IsMonotonic(), res, scope)
	case pmetric.MetricTypeHistogram:
		h := m.Histogram()
		return convertHistogramDataPoints(h.DataPoints(), name, desc, unit, temporalityString(h.AggregationTemporality()), res, scope)
	case pmetric.MetricTypeSummary:
		return convertSummaryDataPoints(m.Summary().DataPoints(), name, desc, unit, res, scope)
	case pmetric.MetricTypeExponentialHistogram:
		eh := m.ExponentialHistogram()
		return convertExpHistogramDataPoints(eh.DataPoints(), name, desc, unit, temporalityString(eh.AggregationTemporality()), res, scope)
	}
	return nil
}

func convertNumberDataPoints(dps pmetric.NumberDataPointSlice, name, desc, unit, typ, temp string, monotonic bool, res store.Resource, scope store.Scope) []store.MetricDataPoint {
	out := make([]store.MetricDataPoint, 0, dps.Len())
	for i := 0; i < dps.Len(); i++ {
		dp := dps.At(i)
		mdp := store.MetricDataPoint{
			Name: name, Description: desc, Unit: unit, Type: typ,
			Timestamp: dp.Timestamp().AsTime(), StartTime: dp.StartTimestamp().AsTime(),
			Attributes: mapToGoMap(dp.Attributes()), Flags: int(dp.Flags()),
			Resource: res, Scope: scope, IsMonotonic: monotonic, Temporality: temp,
		}
		switch dp.ValueType() {
		case pmetric.NumberDataPointValueTypeDouble:
			mdp.Value = dp.DoubleValue()
		case pmetric.NumberDataPointValueTypeInt:
			mdp.Value = float64(dp.IntValue())
		}
		out = append(out, mdp)
	}
	return out
}

func convertHistogramDataPoints(dps pmetric.HistogramDataPointSlice, name, desc, unit, temp string, res store.Resource, scope store.Scope) []store.MetricDataPoint {
	out := make([]store.MetricDataPoint, 0, dps.Len())
	for i := 0; i < dps.Len(); i++ {
		dp := dps.At(i)
		out = append(out, store.MetricDataPoint{
			Name: name, Description: desc, Unit: unit, Type: "histogram",
			Timestamp: dp.Timestamp().AsTime(), StartTime: dp.StartTimestamp().AsTime(),
			Attributes: mapToGoMap(dp.Attributes()), Flags: int(dp.Flags()),
			Resource: res, Scope: scope, Temporality: temp,
			Count: dp.Count(), Sum: dp.Sum(), Min: dp.Min(), Max: dp.Max(),
			BucketCounts:   dp.BucketCounts().AsRaw(),
			ExplicitBounds: dp.ExplicitBounds().AsRaw(),
		})
	}
	return out
}

func convertSummaryDataPoints(dps pmetric.SummaryDataPointSlice, name, desc, unit string, res store.Resource, scope store.Scope) []store.MetricDataPoint {
	out := make([]store.MetricDataPoint, 0, dps.Len())
	for i := 0; i < dps.Len(); i++ {
		dp := dps.At(i)
		qvs := dp.QuantileValues()
		quantiles := make([]store.QuantileValue, 0, qvs.Len())
		for j := 0; j < qvs.Len(); j++ {
			qv := qvs.At(j)
			quantiles = append(quantiles, store.QuantileValue{Quantile: qv.Quantile(), Value: qv.Value()})
		}
		out = append(out, store.MetricDataPoint{
			Name: name, Description: desc, Unit: unit, Type: "summary",
			Timestamp: dp.Timestamp().AsTime(), StartTime: dp.StartTimestamp().AsTime(),
			Attributes: mapToGoMap(dp.Attributes()), Flags: int(dp.Flags()),
			Resource: res, Scope: scope, Count: dp.Count(), Sum: dp.Sum(), Quantiles: quantiles,
		})
	}
	return out
}

func convertExpHistogramDataPoints(dps pmetric.ExponentialHistogramDataPointSlice, name, desc, unit, temp string, res store.Resource, scope store.Scope) []store.MetricDataPoint {
	out := make([]store.MetricDataPoint, 0, dps.Len())
	for i := 0; i < dps.Len(); i++ {
		dp := dps.At(i)
		out = append(out, store.MetricDataPoint{
			Name: name, Description: desc, Unit: unit, Type: "exponential_histogram",
			Timestamp: dp.Timestamp().AsTime(), StartTime: dp.StartTimestamp().AsTime(),
			Attributes: mapToGoMap(dp.Attributes()), Flags: int(dp.Flags()),
			Resource: res, Scope: scope, Temporality: temp,
			Count: dp.Count(), Sum: dp.Sum(), Min: dp.Min(), Max: dp.Max(),
		})
	}
	return out
}

func convertResource(r pcommon.Resource, schemaURL string) store.Resource {
	attrs := mapToGoMap(r.Attributes())
	svcName, _ := attrs["service.name"].(string)
	return store.Resource{ServiceName: svcName, Attributes: attrs, SchemaURL: schemaURL}
}

func convertScope(s pcommon.InstrumentationScope, schemaURL string) store.Scope {
	return store.Scope{Name: s.Name(), Version: s.Version(), SchemaURL: schemaURL}
}

func convertSpanStatus(s ptrace.Status) store.SpanStatus {
	return store.SpanStatus{Code: statusCodeString(s.Code()), Message: s.Message()}
}

func convertEvents(evts ptrace.SpanEventSlice) []store.SpanEvent {
	out := make([]store.SpanEvent, 0, evts.Len())
	for i := 0; i < evts.Len(); i++ {
		e := evts.At(i)
		out = append(out, store.SpanEvent{
			Name:       e.Name(),
			Timestamp:  e.Timestamp().AsTime(),
			Attributes: mapToGoMap(e.Attributes()),
		})
	}
	return out
}

func convertLinks(links ptrace.SpanLinkSlice) []store.SpanLink {
	out := make([]store.SpanLink, 0, links.Len())
	for i := 0; i < links.Len(); i++ {
		l := links.At(i)
		out = append(out, store.SpanLink{
			TraceID:    traceIDStr(l.TraceID()),
			SpanID:     spanIDStr(l.SpanID()),
			Attributes: mapToGoMap(l.Attributes()),
		})
	}
	return out
}

func mapToGoMap(m pcommon.Map) map[string]any {
	if m.Len() == 0 {
		return map[string]any{}
	}
	out := make(map[string]any, m.Len())
	m.Range(func(k string, v pcommon.Value) bool {
		out[k] = valueToGo(v)
		return true
	})
	return out
}

func valueToGo(v pcommon.Value) any {
	switch v.Type() {
	case pcommon.ValueTypeStr:
		return v.Str()
	case pcommon.ValueTypeBool:
		return v.Bool()
	case pcommon.ValueTypeInt:
		return v.Int()
	case pcommon.ValueTypeDouble:
		return v.Double()
	case pcommon.ValueTypeBytes:
		return hex.EncodeToString(v.Bytes().AsRaw())
	case pcommon.ValueTypeSlice:
		sl := v.Slice()
		arr := make([]any, 0, sl.Len())
		for i := 0; i < sl.Len(); i++ {
			arr = append(arr, valueToGo(sl.At(i)))
		}
		return arr
	case pcommon.ValueTypeMap:
		return mapToGoMap(v.Map())
	}
	return nil
}

func valueString(v pcommon.Value) string {
	if v.Type() == pcommon.ValueTypeStr {
		return v.Str()
	}
	return v.AsString()
}

func traceIDStr(id pcommon.TraceID) string {
	b := id
	if b.IsEmpty() {
		return ""
	}
	return hex.EncodeToString(b[:])
}

func spanIDStr(id pcommon.SpanID) string {
	b := id
	if b.IsEmpty() {
		return ""
	}
	return hex.EncodeToString(b[:])
}

func spanKindString(k ptrace.SpanKind) string {
	switch k {
	case ptrace.SpanKindInternal:
		return "INTERNAL"
	case ptrace.SpanKindServer:
		return "SERVER"
	case ptrace.SpanKindClient:
		return "CLIENT"
	case ptrace.SpanKindProducer:
		return "PRODUCER"
	case ptrace.SpanKindConsumer:
		return "CONSUMER"
	default:
		return "UNSPECIFIED"
	}
}

func statusCodeString(c ptrace.StatusCode) string {
	switch c {
	case ptrace.StatusCodeOk:
		return "OK"
	case ptrace.StatusCodeError:
		return "ERROR"
	default:
		return "UNSET"
	}
}

func temporalityString(t pmetric.AggregationTemporality) string {
	switch t {
	case pmetric.AggregationTemporalityDelta:
		return "delta"
	case pmetric.AggregationTemporalityCumulative:
		return "cumulative"
	default:
		return ""
	}
}
